### PR TITLE
LocalModelNode Daemonset Controller Skeleton

### PR DESCRIPTION
**Pull request by HotsauceLee**
**What this PR does / why we need it**: 
Create local model agent daemonset to reconcile the models which should be cached on a given node.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
Controller for reconciling CR defined in https://github.com/kserve/kserve/pull/3978

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

**Review by greenmoon55 — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -137,13 +137,18 @@ manifests: controller-gen
...
+	$(CONTROLLER_GEN) rbac:roleName=kserve-localmodelnode-manager-role paths=./pkg/controller/v1alpha1/localmodelnode output:rbac:artifacts:config=config/rbac/localmodelnode
 	# Copy the cluster role to the helm chart
 	cp config/rbac/auth_proxy_role.yaml charts/kserve-resources/templates/clusterrole.yaml
 	cat config/rbac/role.yaml >> charts/kserve-resources/templates/clusterrole.yaml
 	# Copy the local model role with Helm chart while keeping the Helm template condition
 	echo '{{- if .Values.kserve.localmodel.enabled }}' > charts/kserve-resources/templates/localmodel/role.yaml
 	cat config/rbac/localmodel/role.yaml >> charts/kserve-resources/templates/localmodel/role.yaml
 	echo '{{- end }}' >> charts/kserve-resources/templates/localmodel/role.yaml
+	# Copy the local model node role with Helm chart while keeping the Helm template condition
+	echo '{{- if .Values.kserve.localmodelnode.enabled }}' > charts/kserve-resources/templates/localmodelnode/role.yaml
```
Comment (greenmoon55) [inline Makefile]:
```suggestion
	echo '{{- if .Values.kserve.localmodel.enabled }}'> charts/kserve-resources/templates/localmodelnode/role.yaml
```

File: pkg/controller/v1alpha1/localmodelnode/controller.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +kubebuilder:rbac:groups=serving.kserve.io,resources=inferenceservices,verbs=get;list;watch
+// +kubebuilder:rbac:groups=serving.kserve.io,resources=localmodelnodegroups,verbs=get;list;watch
```
Comment (greenmoon55) [inline pkg/controller/v1alpha1/localmodelnode/controller.go]: Remove

File: pkg/controller/v1alpha1/localmodelnode/controller.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+// +kubebuilder:rbac:groups=serving.kserve.io,resources=clusterstoragecontainers,verbs=get;list;watch
+// +kubebuilder:rbac:groups=serving.kserve.io,resources=localmodelnodes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=serving.kserve.io,resources=localmodelnodes/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=configmaps,verbs=get
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=nodes/status,verbs=get;watch
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=batch,resources=jobs/status,verbs=get
+// +kubebuilder:rbac:groups=core,resources=persistentvolumes,verbs=get;list;watch;create;update;patch;delete
```
Comment (greenmoon55) [inline pkg/controller/v1alpha1/localmodelnode/controller.go]: remove

File: charts/kserve-resources/templates/localmodelnode/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      annotations:
+        kubectl.kubernetes.io/default-container: manager
+    spec:
+      serviceAccountName: kserve-localmodelnode-controller-manager
+      securityContext:
+        runAsNonRoot: true
+      containers:
+      - command:
+        - /manager
+        image: "{{ .Values.kserve.localmodelnode.controller.image }}:{{ .Values.kserve.localmodelnode.controller.tag }}"
```
Comment (greenmoon55) [inline charts/kserve-resources/templates/localmodelnode/deployment.yaml]:
```suggestion
        image: "{{ .Values.kserve.localmodel.daemonsetController.image }}:{{ .Values.kserve.localmodel.daemonsetController.tag }}"
```

File: charts/kserve-resources/templates/localmodelnode/namespace.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{{- if .Values.kserve.localmodel.enabled }}
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: {{.Values.kserve.localmodelnode.jobNamespace}}
```
Comment (greenmoon55) [inline charts/kserve-resources/templates/localmodelnode/namespace.yaml]:
```suggestion
  name: {{.Values.kserve.localmodel.jobNamespace}}
```

File: charts/kserve-resources/templates/localmodelnode/role_binding.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{{- if .Values.kserve.localmodelnode.enabled }}
```
Comment (greenmoon55) [inline charts/kserve-resources/templates/localmodelnode/role_binding.yaml]:
```suggestion
{{- if .Values.kserve.localmodel.enabled }}
```

File: charts/kserve-resources/templates/localmodelnode/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
+{{- if .Values.kserve.localmodel.enabled }}
+---
+apiVersion: apps/v1
+kind: Deployment
```
Comment (greenmoon55) [inline charts/kserve-resources/templates/localmodelnode/deployment.yaml]: Should this be a daemonset?

File: charts/kserve-resources/templates/localmodelnode/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  labels:
+    app.kubernetes.io/name: kserve-localmodelnode-controller-manager
+    control-plane: kserve-localmodelnode-controller-manager
+    controller-tools.k8s.io: "1.0"
+spec:
+  selector:
+    matchLabels:
+      control-plane: kserve-localmodelnode-controller-manager
+      controller-tools.k8s.io: "1.0"
+  template:
```
Comment (greenmoon55) [inline charts/kserve-resources/templates/localmodelnode/deployment.yaml:17]: I think we want to only run the daemonset one on some nodes. e.g. GPU nodes. Then we need to make `matchLabels` configurable in the helm chart. But we can do it in a different PR.

https://kubernetes.io/docs/tasks/manage-daemon/pods-some-nodes/#step-2-create-the-manifest

File: config/default/localmodelnode_manager_image_patch.yaml [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+kind: DaemonSet
+metadata:
+  name: kserve-localmodelnode-controller-manager
+  namespace: kserve
+spec:
+  template:
+    spec:
+      containers:
+      # Change the value of image field below to your controller image URL
+      - image: kserve/kserve-localmodelnode-controller:latest
```
Comment (yuzisun) [inline config/default/localmodelnode_manager_image_patch.yaml]: kserve-localmodelnode-controller -> kserve-localmodelnode-agent

File: charts/kserve-resources/templates/localmodelnode/deployment.yaml (hunk 4 above)
Comment (yuzisun) [inline charts/kserve-resources/templates/localmodelnode/deployment.yaml]:
```suggestion
        image: "{{ .Values.kserve.localmodel.agent.image }}:{{ .Values.kserve.localmodel.agent.tag }}"
```

File: config/default/localmodelnode_manager_image_patch.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: kserve-localmodelnode-controller-manager
```
Comment (yuzisun) [inline config/default/localmodelnode_manager_image_patch.yaml]:
```suggestion
  name: kserve-localmodelnode-agent
```

File: charts/kserve-resources/templates/localmodelnode/deployment.yaml (hunk 8 above)
Comment (yuzisun) [reply charts/kserve-resources/templates/localmodelnode/deployment.yaml:17]: Can we create an issue to track so we do not forget

**Review by yuzisun — APPROVED** [review]
/lgtm
/approve
